test/unit/org/apache/cassandra/db/streaming/EntireSSTableStreamConcurrentComponentMutationTest.java

```diff
@@ -0,0 +1,338 @@
+        store = keyspace.getColumnFamilyStore("Standard1");
+
+        // insert data and compact to a single sstable
+        CompactionManager.instance.disableAutoCompaction();
+        for (int j = 0; j < 10; j++)
+        {
+            new RowUpdateBuilder(store.metadata(), j, String.valueOf(j))
+            .clustering("0")
+            .add("val", ByteBufferUtil.EMPTY_BYTE_BUFFER)
+            .build()
+            .applyUnsafe();
+        }
+        store.forceBlockingFlush();
+        CompactionManager.instance.performMaximal(store, false);
+
+        sstable = store.getLiveSSTables().iterator().next();
+        descriptor = sstable.descriptor;
+
+        Token start = ByteOrderedPartitioner.instance.getTokenFactory().fromString(Long.toHexString(0));
+        Token end = ByteOrderedPartitioner.instance.getTokenFactory().fromString(Long.toHexString(100));
+        rangesAtEndpoint = RangesAtEndpoint.toDummyList(Collections.singleton(new Range<>(start, end)));
+
+        service = Executors.newFixedThreadPool(2);
+    }
+
+    @AfterClass
+    public static void cleanup()
+    {
+        service.shutdown();
+    }
+
+    @After
+    public void reset() throws IOException
+    {
+        latch = new CountDownLatch(1);
+        // reset repair info to avoid test interfering each other
+        descriptor.getMetadataSerializer().mutateRepairMetadata(descriptor, 0, ActiveRepairService.NO_PENDING_REPAIR, false);
+    }
+
+    @Test
+    public void testStream() throws Exception
+    {
+        testStreamWithConcurrentComponentMutation(NO_OP, NO_OP);
+    }
+
+    /**
+     * Entire-sstable-streaming receiver will throw checksum validation failure because concurrent stats metadata
+     * update causes the actual transfered file size to be different from the one in {@link ComponentManifest}
+     */
+    @Test
+    @BMRule(name = "Let component mutation happen immediately after ComponentManifest completion",
+    targetClass = "ComponentManifest",
+    targetMethod = "<init>",
+    action = "Thread.sleep(3000)")
+    public void testStreamWithStatsMutation() throws Exception
+    {
+        testStreamWithConcurrentComponentMutation(NO_OP, () -> {
+
+            Descriptor desc = sstable.descriptor;
+            desc.getMetadataSerializer().mutate(desc, "testing", stats -> stats.mutateRepairedMetadata(0, UUID.randomUUID(), false));
+
+            return null;
+        });
+    }
+
+    @Test
+    @BMRule(name = "Delay saving index summary, manifest may link partially written file if there is no lock",
+            targetClass = "SSTableReader",
+            targetMethod = "saveSummary(Descriptor, DecoratedKey, DecoratedKey, IndexSummary)",
+            targetLocation = "AFTER INVOKE serialize",
+            condition = "$descriptor.cfname.contains(\"Standard1\")",
+            action = "org.apache.cassandra.db.streaming.EntireSSTableStreamConcurrentComponentMutationTest.countDown();Thread.sleep(5000);")
```

Comment:
@jasonstack Awesome. I think this is going to work in most environments. (It certainly works locally for me.) What if, instead of relying on a 5 second wait, we just wait on a separate latch, sort of the opposite of what we do with `countDown()` immediately before? Then we'll suspend the summary saving from finishing while the write continues. When the write is done, we can have it count down on that second latch and free the other thread to finish the save. WDYT?

(I know we've both been bitten by slow testing environments messing with our timing assumptions in the past.)

Reply:
>  Then we'll suspend the summary saving from finishing while the write continues. 

It does reproduce the issue more reliably when there is no lock.. With lock, the test will run into deadlock..